internal/argocd/health.go

```diff
@@ -217,58 +224,123 @@ func (h *applicationHealth) GetApplicationHealth(
 		//
 		// TODO: revisit this when https://github.com/argoproj/argo-cd/pull/18660
 		// 	 is merged and released.
-		cooldown := app.Status.OperationState.FinishedAt.Time.Add(10 * time.Second)
-		if duration := time.Until(cooldown); duration > 0 {
-			time.Sleep(duration)
-
-			// Re-fetch the application to get the latest state.
-			if err := h.argoClient.Get(ctx, key, app); err != nil {
-				err = fmt.Errorf("error finding Argo CD Application %q in namespace %q: %w", key.Name, key.Namespace, err)
-				if client.IgnoreNotFound(err) == nil {
-					err = fmt.Errorf("unable to find Argo CD Application %q in namespace %q", key.Name, key.Namespace)
+		if app.Status.OperationState != nil {
+			cooldown := app.Status.OperationState.FinishedAt.Time.Add(10 * time.Second)
+			if duration := time.Until(cooldown); duration > 0 {
+				time.Sleep(duration)
+
+				// Re-fetch the application to get the latest state.
+				if err := h.argoClient.Get(ctx, key, app); err != nil {
+					err = fmt.Errorf("error finding Argo CD Application %q in namespace %q: %w", key.Name, key.Namespace, err)
+					if client.IgnoreNotFound(err) == nil {
+						err = fmt.Errorf("unable to find Argo CD Application %q in namespace %q", key.Name, key.Namespace)
+					}
+					return kargoapi.HealthStateUnknown, healthStatus, syncStatus, err
 				}
-				return kargoapi.HealthStateUnknown, healthStatus, syncStatus, err
 			}
 		}
 	}
 
 	// With all the above checks passed, we can now assume the Argo CD
 	// Application's health state is reliable.
-	healthState, err := stageHealthForAppHealth(app)
+	healthState, err := stageHealthForAppHealth(ctx, app)
 	return healthState, healthStatus, syncStatus, err
 }
 
 // stageHealthForAppSync returns the v1alpha1.HealthState for an Argo CD
 // Application based on its sync status.
-func stageHealthForAppSync(app *argocd.Application, revision string) (kargoapi.HealthState, error) {
+func stageHealthForAppSync(
+	ctx context.Context,
+	app *argocd.Application,
+	revisions []string) (kargoapi.HealthState, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues("appName", app.GetName(), "revisions", revisions)
+	logger.Debug("About to determine stage health based on app sync status.")
 	switch {
-	case revision == "":
+	case revisions == nil || (len(revisions) == 1 && revisions[0] == ""):
```

Comment:
Another spot where the nil check is not needed.

Reply:
Also this seems to treat _one_ empty string "desired revision" as a special case, but I think the special case should _actually_ be "_all_ desired revisions are empty strings."

It may be that the best way to handle that is when _finding_ the desired revisions. If we come up empty-handed for each and every source, we should probably just return nil/empty slice so that everywhere else we don't need to have the logic that checks the whole collection to see if its all empties.